src/SnapKeyring.ts

```diff
@@ -436,6 +436,21 @@ export class SnapKeyring extends EventEmitter {
     // If the snap answers asynchronously, we will inform the user with a redirect
     if (response.redirect?.message || response.redirect?.url) {
       const { message = '', url = '' } = response.redirect;
+
+      // Check redirect url domain is in the snap allowed origins
+      if (url) {
+        const { origin } = new URL(url);
+        const snap = this.#snapClient.getController().get(snapId);
+        const allowedOrigins =
+          snap?.manifest?.initialPermissions['endowment:keyring']
+            ?.allowedOrigins ?? [];
+        if (!allowedOrigins.includes(origin)) {
+          throw new Error(
+            `Redirect URL domain '${origin}' is not an allowed origin by snap '${snapId}'`,
+          );
+        }
+      }
```

Comment:
Looking at this piece of code I had a couple of thoughts, let me know what you think,

1. We could have a method to provide the `allowedOrigins` like `getSnapAllowedOrigins(snap: Snap): string[]`.
2. About this chaining `snap?.manifest?.initialPermissions['endowment:keyring']?.allowedOrigins`. Do you think returning an empty list in case some of these values is `undefined` is the correct way of handling it? I imagine that we expect those values to always be defined, at least until `initialPermissions`, I'm wondering if a more accurate error should be throw in that case. Also, have you seen a scenario where any of those could be `undefined` and it's acceptable?
3. Lastly, this method (`#submitRequest`) is getting quite bloated, maybe it would be good to extract this logic into it's own method

Reply:
I agree with points 1 and 3, but I'm ok with the `... ? []` approach, it looks cleaner IMHO and is still functionally correct.

Reply:
1. Very good idea!
2. I agree that I do not see any instance where things would be undefined up until the `initialPermissions` as these are required from the returned object from the controller (if the snap ID look-up does not come up undefined). There is not a scenario where I could think it possible/allowed for any of these properties to be undefined if the snap itself is defined.

I addressed the utility method in this commit [43d7d61](https://github.com/MetaMask/eth-snap-keyring/pull/228/commits/43d7d617b3583a3578843d626d3efb21fcfdb441). I added throwing an error if the snap does not have allowed origins. I updated the tests in this commit [61302ba](https://github.com/MetaMask/eth-snap-keyring/pull/228/commits/61302ba462facb9c4ccdf268e4579549579bc6e3) and added a check that the snap itself is not undefined.

Reply:
>  I'm ok with the `... ? []` approach, it looks cleaner IMHO and is still functionally correct.

I do not have a strong opinion but reverted the change to throw an error and instead return an empty `[]` in this commit [29a1904](https://github.com/MetaMask/eth-snap-keyring/pull/228/commits/29a1904a3b9c918781f3c09513e44fe82dcbab38)

Reply:
@gantunesr and @danroc 
To address comment point #3 above I made another [PR](https://github.com/MetaMask/eth-snap-keyring/pull/232) to extract the logic within `#submitRequest`